enriched-span-constants/src/main/java/org/hypertrace/traceenricher/enrichedspan/constants/utils/EnrichedSpanUtils.java

```diff
@@ -301,12 +317,74 @@ public static String getUserAgent(Event event) {
     return Optional.empty();
   }
 
+  /**
+   * This method returns an Optional which, if present, contains the full URL, including the protocol and host name.
+   * If the tags in the raw span do not contain the full URL, it tries to construct a full URL using the protocol
+   * ("http") and host name.
+   */
   public static Optional<String> getFullHttpUrl(Event event) {
-    if (event.getHttp() != null && event.getHttp().getRequest() != null) {
-      return Optional.ofNullable(event.getHttp().getRequest().getUrl());
+    Optional<String> optionalUrl = Optional.ofNullable(event.getHttp())
+            .map(org.hypertrace.core.datamodel.eventfields.http.Http::getRequest)
+            .map(Request::getUrl);
+
+    if (optionalUrl.isEmpty() && event.getAttributes() != null && event.getAttributes().getAttributeMap() != null) {
+      // iterate on all URL related attribute keys and find / create one with protocol, host header and path
+      Map<String, AttributeValue> attributesMap = event.getAttributes().getAttributeMap();
+      for (String urlAttribute : FULL_URL_ATTRIBUTES) {
+        if (attributesMap.containsKey(urlAttribute)) {
+          String url = attributesMap.get(urlAttribute).getValue();
+          if (url != null && url.toLowerCase().startsWith(HTTP_PREFIX)) {
+            optionalUrl = Optional.of(url);
+          } else {
+            optionalUrl = constructHttpUrlFromHostName(event);
+          }
+        }
+      }
     }
+    return optionalUrl;
+  }
+
+  private static Optional<String> constructHttpUrlFromHostName(Event event) {
+    Optional<String> optionalUrl = Optional.empty();
+    Optional<RequestHeaders> requestHeaders = Optional.ofNullable(event.getHttp())
+            .map(org.hypertrace.core.datamodel.eventfields.http.Http::getRequest)
+            .map(Request::getHeaders);
+    Optional<String> hostName = requestHeaders.map(RequestHeaders::getHost);
```

Comment:
@GurtejSohi a better way to read host header is from the `HOST_HEADER` enriched attribute. That would have handled all corner cases to get host header.